### PR TITLE
Add support for JoinPoolNoSwap message to Wasm Bindings

### PR DESCRIPTION
**Pull request by georgemc98**
## What is the purpose of the change

This PR adds support for the JoinPoolMsg.  The `osmo_reflect.wasm` file necessary for running tests are dependent on this PR in the `osmosis-bindings` repo: https://github.com/osmosis-labs/osmosis-bindings/pull/8

> Add a description of the overall background and high level changes that this PR introduces

This pull request adds support for Cosmwasm smart contracts to join a liquidity pool.

## Brief Changelog
 
- Adds JoinPool to the Wasm bindings
- Adds unit test to check errors and logic
- Compiled the `osmo_reflect.wasm` file necessary for unit tests (This is dependent on this PR:  https://github.com/osmosis-labs/osmosis-bindings/pull/8)

## Testing and Verifying

This change added tests and can be verified as follows:

  - Added unit test that validates the the correct amount of shares are minted when a user joins a liquidity pool
  - Added unit test that validates the LP token is minted back to the contract
  
## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? yes
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? no
  - How is the feature or change documented? The changes provide a wrapper around the function: (x/gamm/keeper/pool_service/JoinPoolNoSwap)

### CODE REVIEW

File: app/wasm/bindings/msg.go [hunk 1]
```diff
@@ -33,3 +36,9 @@ type SwapMsg struct {
 	Route  []Step              `json:"route"`
 	Amount SwapAmountWithLimit `json:"amount"`
 }
+
+type JoinPool struct {
```
Comment (ValarDragon) [inline app/wasm/bindings/msg.go:40]: Can we make this JoinPoolNoSwap ?

Comment (georgemc98) [reply app/wasm/bindings/msg.go:40]: For sure.  I can switch this rn.

Comment (georgemc98) [reply app/wasm/bindings/msg.go]: Done

File: app/wasm/test/custom_msg_test.go [hunk 2]
```diff
@@ -514,6 +515,98 @@ func TestSwapMsg(t *testing.T) {
 	}
 }
 
+func TestJoinPoolMsg(t *testing.T) {
```
Comment (p0mvn) [inline app/wasm/test/custom_msg_test.go:518]:
```suggestion
func TestJoinPoolNoSwapMsg(t *testing.T) {
```

File: app/wasm/test/custom_msg_test.go [hunk 3]
```diff
@@ -514,6 +515,98 @@ func TestSwapMsg(t *testing.T) {
...
+	err = executeCustom(t, ctx, osmosis, reflect, provider, msg, sdk.NewCoins(sdk.NewCoin("random", sdk.NewInt(10))))
+
+	require.Error(t, err)
+	require.Equal(t, "0random is smaller than 10random: insufficient funds", err.Error())
+
+	err = executeCustom(t, ctx, osmosis, reflect, provider, msg, osmoStarLiquidity)
+
+	require.NoError(t, err)
+
+	//Simualate sending share tokens to user
```
Comment (p0mvn) [inline app/wasm/test/custom_msg_test.go:589]: Just trying to understand - how does everything below help test the `JoinPoolNoSwap` message?

This seems like an integration level test if the goal is to ensure that components work well together

File: app/wasm/test/messages_test.go [hunk 4]
```diff
@@ -593,3 +594,80 @@ func TestSwapMultiHop(t *testing.T) {
...
+		sdk.NewInt64Coin("ustar", 240_000_000),
+	}
+	// 20 star to 1 osmo
+	starPool := preparePool(t, ctx, osmosis, actor, poolFunds)
+
+	osmoStarLiquidity := sdk.NewCoins(sdk.NewInt64Coin("uosmo", 12_000),
+		sdk.NewInt64Coin("ustar", 240_000))
+
+	specs := map[string]struct {
+		join_pool_no_swap *wasmbindings.JoinPoolNoSwap
```
[the remark this answers is not in the document]
Comment (georgemc98) [reply app/wasm/test/messages_test.go:615]: Got it.  I've changed this.

File: app/wasm/test/messages_test.go [hunk 5]
```diff
@@ -593,3 +594,80 @@ func TestSwapMultiHop(t *testing.T) {
...
+		expErr            bool
```
[the remark this answers is not in the document]
Comment (georgemc98) [reply app/wasm/test/messages_test.go:616]: Got it.  I can do this today.

File: app/wasm/test/custom_msg_test.go (hunk 3 above)
Comment (georgemc98) [reply app/wasm/test/custom_msg_test.go:589]: Good point.  I can make the change to ensure the shareToken is sent to the `reflect` address, not the `provider` address.  I wanted to ensure that the share and liquidity is updated correctly.

File: app/wasm/message_plugin.go [hunk 6]
```diff
@@ -194,6 +197,28 @@ func PerformSwap(keeper *gammkeeper.Keeper, ctx sdk.Context, contractAddr sdk.Ac
 	}
 }
 
+func (m *CustomMessenger) joinPool(ctx sdk.Context, contractAddr sdk.AccAddress, joinPool *wasmbindings.JoinPoolNoSwap) ([]sdk.Event, [][]byte, error) {
+	err := PerformJoin(m.gammKeeper, ctx, contractAddr, joinPool)
```
[the remark this answers is not in the document]
Comment (p0mvn) [reply app/wasm/message_plugin.go:201]: I think it's fine. I realized that this pattern was used for all other messages in this file after I originally commented.

We can refactor all of them at once in a separate PR if needed.

I'll resolve the comment. Sorry for the confusion.

File: app/wasm/test/messages_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (georgemc98) [reply app/wasm/test/messages_test.go]: My latest commit explicitly checks the returned errors.

**Comment by georgemc98** [comment]
Closed in favor of (https://github.com/osmosis-labs/osmosis/pull/1647) which provides more functionality

**Comment by p0mvn** [comment]
@georgemc98 have you found the answer to your question?

I'm not sure, and I was going to suggest to go with either and refactor later if needed

> @p0mvn @ValarDragon Quick question. Should the `PerformJoin` function send the LP token back to the sender of the message? I noticed that the `PerformSwap` function does not send the token back to the sender of the message, whereas the `PerformMint` function does send the newly minted token back to the sender of the message.

**Comment by georgemc98** [comment]
> @georgemc98 have you found the answer to your question?
> 
> I'm not sure, and I was going to suggest to go with either and refactor later if needed
> 
> > @p0mvn @ValarDragon Quick question. Should the `PerformJoin` function send the LP token back to the sender of the message? I noticed that the `PerformSwap` function does not send the token back to the sender of the message, whereas the `PerformMint` function does send the newly minted token back to the sender of the message.

I'm keeping it as is and following the `PerformSwap` logic and not sending the LP token back to sender.

**Pull request closed without merging**